src/ops/array.rs

```diff
@@ -94,32 +98,35 @@ pub(crate) fn add_by_internal(value: &mut [u32], by: &[u32]) -> u32 {
 }
 
 pub(crate) fn add_by_internal_flattened(value: &mut [u32; 3], by: u32) -> u32 {
-    let mut carry: u64;
-    let mut sum: u64;
-    sum = u64::from(value[0]) + u64::from(by);
-    value[0] = (sum & U32_MASK) as u32;
-    carry = sum >> 32;
-    if carry > 0 {
-        sum = u64::from(value[1]) + carry;
-        value[1] = (sum & U32_MASK) as u32;
-        carry = sum >> 32;
-        if carry > 0 {
-            sum = u64::from(value[2]) + carry;
-            value[2] = (sum & U32_MASK) as u32;
-            carry = sum >> 32;
-        }
-    }
-    carry as u32
+    manage_add_by_internal(by, value)
 }
 
 #[inline]
-pub(crate) fn add_one_internal(value: &mut [u32]) -> u32 {
-    let mut carry: u64 = 1; // Start with one, since adding one
-    let mut sum: u64;
-    for i in value.iter_mut() {
-        sum = (*i as u64) + carry;
-        *i = (sum & U32_MASK) as u32;
-        carry = sum >> 32;
+pub(crate) fn add_one_internal(value: &mut [u32; 3]) -> u32 {
+    manage_add_by_internal(1, value)
+}
+
+// `u64 as u32` are safe because of widening and 32bits shifts
+#[inline]
+pub(crate) fn manage_add_by_internal<const N: usize>(initial_carry: u32, value: &mut [u32; N]) -> u32 {
+    let mut carry = u64::from(initial_carry);
+    let mut iter = 0..value.len();
+    let mut sum = 0;
+
+    let mut sum_fn = |local_carry: &mut u64, idx| {
+        sum = u64::from(value[idx]).wrapping_add(*local_carry);
+        value[idx] = (sum & U32_MASK) as u32;
+        *local_carry = sum.wrapping_shr(32);
+    };
+
+    if let Some(idx) = iter.next() {
+        sum_fn(&mut carry, idx);
+    }
+
+    for idx in iter {
+        if carry > 0 {
+            sum_fn(&mut carry, idx);
+        }
     }
```

Comment:
https://godbolt.org/z/qK5qzEK81

Both old and new versions have the same assembly output (taking aside an initial variable), i.e., the compiler is unrolling the loop. IMO, this new version is much more readable and flexible for any `N`.

cc @chris-cantor @TradingTomatillo